src/main/clojure/osgi/deps.clj

```diff
@@ -0,0 +1,390 @@
+
+(defn package-deps []
+  (let [bundles (->> (osgi/bundle-list)
+                     (map osgi/bundle->map))
+        deps (get-deps bundles)]
+    (dot/digraph
+      [;(map #(-> [(:id %) {:label (:name %) :shape :box}]) bundles)
+       (->> bundles
+            (filter #(re-matches #"catalog-ui-search" (:name %)))
+            (mapcat
+              (fn [bundle]
+                (map #(-> [(:name bundle) (get deps (first %))])
+                     (get-in bundle [:headers :import-package])))))])))
+
+(defn service-deps []
+  (let [bundles (map osgi/bundle->map (take-last 10 (osgi/bundle-list)))]
+    (dot/digraph
+      [(map #(-> [(:id %) {:label (:name %) :shape :box}]) bundles)])))
+
+(defn
+  feature->map
+  [f] {
+       :id            (.getId f)
+       :name          (.getName f)
+       :description   (.getDescription f)
+       :details       (.getDetails f)
+       :version       (.getVersion f)
+       :resolver      (.getResolver f)
+       :install       (.getInstall f)
+       :isHidden      (.isHidden f)
+       :dependencies  (map (fn [d] {
+                                    :name         (.getName d)
+                                    :version      (.getVersion d)
+                                    :isPrereq     (.isPrerequisite d)
+                                    :isDependency (.isDependency d)})
+                           (.getDependencies f))
+       :bundles       (map (fn [b] {
+                                    :location         (.getLocation b)
+                                    :originalLocation (.getOriginalLocation b)
+                                    :startLevel       (.getStartLevel b)
+                                    :isStart          (.isStart b)
+                                    :isDependency     (.isDependency b)})
+                           (.getBundles f))
+       :startLevel    (.getStartLevel f)
+       :namespace     (.getNamespace f)
+       :resourceRepos (.getResourceRepositories f)
+       :repoUrl       (.getRepositoryUrl f)})
+
+; ##################################################################
+
+(defn bundles [f]
+  (->> (osgi/bundle-list)
+       (map osgi/bundle->map)
+       (filter f)))
+
+(defn features []
+  (->> (osgi/list-features)
+       (map feature->map)))
+
+(comment
+  (->> (bundles (fn [x] true))
+       (filter #(.contains (:name %) "directorymonitor"))
+       (first)))
+
+(comment
+  (->> (features)
+       (filter #(.contains (:name %) "directorymonitor"))
+       (first)))
+
+; ##################################################################
+
+(defn bundles-partitioned
+  "Example function that partitions bundles. Takes a
+  coll of bundle defs and returns a coll of bundle def colls."
+  [bundles]
+  (let [cushion 1
+        id-set (set (map :id bundles))]
+    (->> bundles
+         (partition-by
+           #(and
+              (contains? id-set (+ (:id %) cushion))
+              (contains? id-set (- (:id %) cushion)))))))
```

Comment:
✏️ Still working on a better way to express partitioning in the final graph definition itself.